File: src/atlasforge/models/structure.py
# SPDX-FileCopyrightText: 2026 Dong Lab, Yale School of Medicine <https://donglab.org>
#
# SPDX-License-Identifier: Apache-2.0

from typing import Literal

from pydantic import BaseModel

FeatureType = Literal[
    "transmembrane",
    "intramembrane",
    "topological_domain",
    "binding_site",
    "active_site",
    "glycosylation",
    "disulfide_bond",
    "signal_peptide",
]


class Structure(BaseModel):
    gene_id: str
    symbol: str | None = None
    uniprot_accession: str | None = None
    uniprot_id: str | None = None
    uniprot_length: int | None = None
    # Agreement with the Ensembl canonical protein: exact, isoform, differs, or unknown
    seq_agreement: str | None = None
    id_route: str | None = None
    afdb_entry_id: str | None = None
    afdb_version: int | None = None
    model_file: str | None = None
    model_format: str | None = None
    model_available: bool = False
    model_source_url: str | None = None
    model_page_url: str | None = None
    mean_plddt: float | None = None
    frac_plddt_very_high: float | None = None
    frac_plddt_confident: float | None = None
    frac_plddt_low: float | None = None
    frac_plddt_very_low: float | None = None
    model_created: str | None = None
    model_is_canonical: bool = False
    alphafill_page_url: str | None = None
    n_transmembrane: int = 0
    n_intramembrane: int = 0
    n_binding_sites: int = 0
    n_binding_residues: int = 0
    n_experimental: int = 0
    best_pdb_id: str | None = None
    best_method: str | None = None
    best_resolution: float | None = None


class StructureDetail(Structure):
    """One gene, with the per-residue columns the topology figure draws: AlphaFold's
    confidence, and the canonical sequence the snake plot letters its beads with."""

    plddt: list[int] | None = None
    sequence: str | None = None


class MembraneSegment(BaseModel):
    start: int
    end: int
    kind: Literal["transmembrane", "intramembrane"]


class FeatureSpan(BaseModel):
    """One drawable mark outside the membrane, which is all the overview needs of it."""

    start: int
    end: int
    kind: Literal[
        "binding_site",
        "active_site",
        "glycosylation",
        "disulfide_bond",
        "signal_peptide",
    ]


class GeneTopology(BaseModel):
    gene_id: str
    uniprot_length: int | None = None
    segments: list[MembraneSegment] = []
    features: list[FeatureSpan] = []


class ProteinFeature(BaseModel):
    gene_id: str
    uniprot_accession: str | None = None
    feature_type: FeatureType
    start: int
    end: int
    description: str | None = None
    ligand_name: str | None = None
    ligand_chebi: str | None = None
    ligand_label: int | None = None


class ExperimentalStructure(BaseModel):
    gene_id: str
    uniprot_accession: str | None = None
    pdb_id: str
    method: str | None = None
    resolution: float | None = None
    coverage: float | None = None
    uniprot_start: int | None = None
    uniprot_end: int | None = None
    chains: str | None = None
    ligand_ccd: str | None = None
    model_url: str | None = None
    model_page_url: str | None = None
    created: str | None = None


class DataSourceRecord(BaseModel):
    source: str
    domain: str | None = None
    version: str | None = None
    assembly: str | None = None
    retrieved_date: str | None = None
    citation: str | None = None
    citation_url: str | None = None
    license_spdx: str | None = None
    url: str | None = None

File: web/src/types/structure.ts
// SPDX-FileCopyrightText: 2026 Dong Lab, Yale School of Medicine <https://donglab.org>
//
// SPDX-License-Identifier: Apache-2.0

export interface StructureRecord {
  gene_id: string
  symbol: string | null
  uniprot_accession: string | null
  uniprot_id: string | null
  uniprot_length: number | null
  seq_agreement: SeqAgreement | null
  id_route: string | null
  afdb_entry_id: string | null
  afdb_version: number | null
  model_file: string | null
  model_format: string | null
  model_available: boolean
  model_source_url: string | null
  model_page_url: string | null
  mean_plddt: number | null
  frac_plddt_very_high: number | null
  frac_plddt_confident: number | null
  frac_plddt_low: number | null
  frac_plddt_very_low: number | null
  model_created: string | null
  model_is_canonical: boolean
  alphafill_page_url: string | null
  n_transmembrane: number
  n_intramembrane: number
  n_binding_sites: number
  n_binding_residues: number
  n_experimental: number
  best_pdb_id: string | null
  best_method: string | null
  best_resolution: number | null
}

export interface StructureDetail extends StructureRecord {
  plddt: number[] | null
  sequence: string | null
}

export type SeqAgreement = "exact" | "isoform" | "differs" | "unknown"

export type FeatureType =
  | "transmembrane"
  | "intramembrane"
  | "topological_domain"
  | "binding_site"
  | "active_site"
  | "glycosylation"
  | "disulfide_bond"
  | "signal_peptide"

export type DrawnFeature = Exclude<
  FeatureType,
  "transmembrane" | "intramembrane" | "topological_domain"
>

export interface MembraneSegment {
  start: number
  end: number
  kind: "transmembrane" | "intramembrane"
}

export interface FeatureSpan {
  start: number
  end: number
  kind: DrawnFeature
}

export interface GeneTopology {
  gene_id: string
  uniprot_length: number | null
  segments: MembraneSegment[]
  features: FeatureSpan[]
}

export interface ProteinFeature {
  gene_id: string
  uniprot_accession: string | null
  feature_type: FeatureType
  start: number
  end: number
  description: string | null
  ligand_name: string | null
  ligand_chebi: string | null
  // Distinguishes separate sites of the same ligand, e.g. Na(+) site 1 vs site 2
  ligand_label: number | null
}

export interface ExperimentalStructure {
  gene_id: string
  uniprot_accession: string | null
  pdb_id: string
  method: string | null
  resolution: number | null
  coverage: number | null
  uniprot_start: number | null
  uniprot_end: number | null
  chains: string | null
  ligand_ccd: string | null
  model_url: string | null
  model_page_url: string | null
  created: string | null
}

export interface DataSourceRecord {
  source: string
  domain: string | null
  version: string | null
  assembly: string | null
  retrieved_date: string | null
  citation: string | null
  citation_url: string | null
  license_spdx: string | null
  url: string | null
}
